### PR TITLE
chore: readd browser setup hook

### PR DESCRIPTION
**Pull request by Lightning00Blade**
There are quite a few changes here will comment on the important ones.

### CODE REVIEW

File: test/src/chromiumonly.spec.ts [hunk 1]
```diff
@@ -124,13 +124,19 @@ describe('Chromium-Specific Launcher tests', function () {
...
     });
-    it('should fire "disconnected" when closing with pipe', async () => {
+    it('should fire "disconnected" when closing with pipe', async function () {
       const {browser, close} = await launch({pipe: true});
       try {
         const disconnectedEventPromise = waitEvent(browser, 'disconnected');
         // Emulate user exiting browser.
         browser.process()!.kill();
-        await disconnectedEventPromise;
+        await Deferred.race([
```
Comment (Lightning00Blade) [inline test/src/chromiumonly.spec.ts:133]: Use Puppeteer Deferred to Clear Timeouts automagically

**Review by OrKoN — APPROVED** [review]

File: test/src/headful.spec.ts [hunk 2]
```diff
@@ -35,7 +35,7 @@ const serviceWorkerExtensionPath = path.join(
   'extension'
 );
 
-describe('headful tests', function () {
+describe.skip('headful tests', function () {
```
Comment (Lightning00Blade) [inline test/src/headful.spec.ts]:
```suggestion
describe('headful tests', function () {
```

File: test/src/launcher.spec.ts [hunk 3]
```diff
@@ -562,7 +562,7 @@ describe('Launcher specs', function () {
         });
         expect(error.message).toContain('either pipe or debugging port');
       });
-      it('should launch Chrome properly with --no-startup-window and waitForInitialPage=false', async () => {
+      it.skip('should launch Chrome properly with --no-startup-window and waitForInitialPage=false', async () => {
```
Comment (Lightning00Blade) [inline test/src/launcher.spec.ts]:
```suggestion
      it('should launch Chrome properly with --no-startup-window and waitForInitialPage=false', async () => {
```

File: test/src/mocha-utils.ts [hunk 4]
```diff
@@ -143,6 +145,23 @@ const setupServer = async () => {
   return {server, httpsServer};
 };
 
+export const setupTestBrowserHooks = (): void => {
+  before(async function () {
+    this.timeout(0);
+    try {
+      if (!state.browser) {
+        state.browser = await puppeteer.launch({
```
Comment (Lightning00Blade) [inline test/src/mocha-utils.ts:143]: Create the browser once, and reuse it as you please.

File: test/src/mocha-utils.ts [hunk 5]
```diff
@@ -389,30 +431,34 @@ export const createTimeout = <T>(
   });
 };
 
-let browserCleanups: Array<() => Promise<void>> = [];
+const browserCleanupsAfterAll: Array<() => Promise<void>> = [];
```
Comment (Lightning00Blade) [inline test/src/mocha-utils.ts:424]: Some test need to run the Cleanup after all the test so they don't spawn browser for each test;

File: test/src/stacktrace.spec.ts [hunk 6]
```diff
@@ -18,12 +18,19 @@ import assert from 'assert';
 
 import expect from 'expect';
 
-import {getTestState} from './mocha-utils.js';
+import {getTestState, setupTestBrowserHooks} from './mocha-utils.js';
 import {waitEvent} from './utils.js';
 
 const FILENAME = __filename.replace(/[/\-\\^$*+?.()|[\]{}]/g, '\\$&');
+const parseStackTrace = (stack: string): string => {
```
Comment (Lightning00Blade) [inline test/src/stacktrace.spec.ts:25]: Makes this test less reliant on the file as adding a new line somewhere will cause it to brake;
@jrandolf PTAL

File: tools/mochaRunner/src/utils.ts [hunk 7]
```diff
@@ -171,6 +165,15 @@ export function getExpectationUpdates(
     // If an error occurs during a hook
     // the error not have a file associated with it
     if (!failure.file) {
+      addEntry({
```
Comment (Lightning00Blade) [inline tools/mochaRunner/src/utils.ts:168]: Add a Hook entry the hooks should have logged the correct file that it happen.

File: test/src/mocha-utils.ts [hunk 8]
```diff
@@ -143,6 +144,22 @@ const setupServer = async () => {
   return {server, httpsServer};
 };
 
+export const setupTestBrowserHooks = (): void => {
+  before(async function () {
+    this.timeout(0);
```
Comment (OrKoN) [inline test/src/mocha-utils.ts]: what does timeout(0) do?

File: test/src/mocha-utils.ts (hunk 5 above)
Comment (OrKoN) [reply test/src/mocha-utils.ts:424]: hm, I am not following, could you be more specific about which tests need it?

File: tools/mochaRunner/src/utils.ts (hunk 7 above)
Comment (OrKoN) [reply tools/mochaRunner/src/utils.ts:168]: Was not able to parse the sentence :D

Comment (Lightning00Blade) [reply tools/mochaRunner/src/utils.ts:168]: Neither can I :no_mouth: . We add and entry rather then skipping as we want to investigate the failer. If the error comes from our code we should see the file name in the error message

File: test/src/mocha-utils.ts (hunk 5 above)
Comment (Lightning00Blade) [reply test/src/mocha-utils.ts:424]: `ignoreHTTPSErrors` and `OOPIF` Test set-up their own browsers once.
If we don't split them we get error that the browser has not been closed after each test.

File: test/src/mocha-utils.ts (hunk 8 above)
Comment (Lightning00Blade) [reply test/src/mocha-utils.ts]: Stops the timeout. As discussed offline we will change to use the puppeteer.launch timeout option set to lower that the deafult.

File: test/src/mocha-utils.ts [hunk 9]
```diff
@@ -389,30 +429,34 @@ export const createTimeout = <T>(
...
+const closeLaunched = (storage: Array<() => Promise<void>>) => {
+  return async () => {
+    let cleanup = storage.pop();
+    try {
+      while (cleanup) {
+        await cleanup();
+        cleanup = storage.pop();
+      }
+    } catch (error) {
+      // If the browser was closed by other mean swallow the error
```
Comment (OrKoN) [inline test/src/mocha-utils.ts]:
```suggestion
      // If the browser was closed by other means, swallow the error
```

File: test/src/mocha-utils.ts [hunk 10]
```diff
@@ -389,30 +429,34 @@ export const createTimeout = <T>(
...
+      // and mark he browser as closed
```
Comment (OrKoN) [inline test/src/mocha-utils.ts]:
```suggestion
      // and mark the browser as closed.
```

File: test/src/mocha-utils.ts [hunk 11]
```diff
@@ -389,30 +429,34 @@ export const createTimeout = <T>(
...
+      if ((error as any)?.message.includes('Connection closed')) {
```
Comment (OrKoN) [inline test/src/mocha-utils.ts]:
```suggestion
      if ((error as Error)?.message.includes('Connection closed')) {
```

**Review by OrKoN — APPROVED** [review]

**Pull request merged**